src/ucs/memory/numa.c

```diff
@@ -72,3 +86,179 @@ int ucs_numa_node_of_cpu(int cpu)
+{
+    const uint8_t shift = sizeof(ucs_numa_node_t) * 8;
+
+    return (ucs_max(node1, node2) << shift) | ucs_min(node1, node2);
+}
+
+/**
+ * Iterate the directory entries in @param path
+ * with the form @param prefix<index> and return the
+ * maximum index.
+ * <index> is a positive number.
+ *
+ * @param [in]  path Directory path
+ * @param [in]  prefix Consider only entries starting with prefix
+ * @param [in]  prefix_length Prefix length
+ * @param [in]  max index range is [0, max]
+ *
+ * @return Maximum index of the entries starting with prefix.
+ */
+static int ucs_numa_get_max_dirent(const char *path, const char *prefix,
+                                   const size_t prefix_length,
+                                   const uint32_t max)
+{
+    int max_index = 0;
+    DIR *dir;
+    struct dirent *de;
+    int idx;
+
+    dir = opendir(path);
+    if (!dir) {
+        return -1;
+    } else {
+        while ((de = readdir(dir)) != NULL) {
+            if (strncmp(de->d_name, prefix, prefix_length)) {
+                continue;
+            }
+
+            idx = strtoul(de->d_name + prefix_length, 0, 0);
+            ucs_assert(idx >= 0);
+            ucs_assert(idx <= max);
+            if (max_index < idx) {
+                max_index = idx;
+            }
+        }
+
+        closedir(dir);
+    }
+
+    return max_index;
+}
+
+int ucs_numa_get_max_node()
+{
+    static int max_node = -1;
+
+    if (max_node == -1) {
+        max_node = ucs_numa_get_max_dirent(UCS_NUMA_NODES_DIR_PATH, "node", 4,
+                                           UCS_NUMA_MAX_NODE);
+        if (max_node == -1) {
+            /* TODO - Handle Error */
+        }
+    }
+
+    return max_node;
+}
+
+int ucs_numa_num_configured_cpus()
+{
+    static int max_cpu = -1;
+
+    if (max_cpu == -1) {
+        max_cpu = ucs_numa_get_max_dirent(UCS_NUMA_CORES_DIR_PATH, "cpu", 3,
+                                          CPU_SETSIZE);
+        if (max_cpu == -1) {
+            /* TODO - Handle Error */
+        }
+
+        max_cpu += 1;
+    }
+
+    return max_cpu;
+}
+
+ucs_numa_node_t ucs_numa_node_of_cpu_v2(int cpu)
+{
+    /* we can initialize statically only to the value 0, so the NUMA node
+     * numbers will be stored as 1..N instead of 0..N-1 */
+    ucs_numa_node_t node                                = 0;
+    static ucs_numa_node_t cpu_numa_node[__CPU_SETSIZE] = {0};
+    char core_dir_path[PATH_MAX];
+
+    ucs_assert(cpu < __CPU_SETSIZE);
+
+    if (cpu_numa_node[cpu] == 0) {
+        ucs_snprintf_safe(core_dir_path, PATH_MAX, UCS_NUMA_CORE_DIR_PATH, cpu);
+        node               = ucs_numa_get_max_dirent(core_dir_path, "node", 4,
+                                                     ucs_numa_get_max_node());
+        cpu_numa_node[cpu] = node + 1;
+    }
+
+    return cpu_numa_node[cpu] - 1;
+}
+
+static ucs_numa_distance_t
+ucs_numa_node_get_distance(char *distances, ucs_numa_node_t dest)
```

Comment:
Because we parsing the distances string we can populate the hash with distances from source to all other nodes, but because we are sorting the key it will require querying the hash if the element already exists or preserve ordering when adding to the hash (adding all the distances from the higher idx node).